### ISSUE TITLE
Chapter 13: Error in Submitting comments. Example 13-4

### ISSUE THREAD

**Issue by EryanNeo** [miguelgrinberg/flasky#563]
While submitting comments on a post, im getting an error of:

Build Error: Could not build url for endpoint 'main.post' with values ['page']. Did you forget to specify values ['id'].

Could you please clarify what might be the issue.

**Comment by miguelgrinberg** [comment]
It means that a call to `url_for()` is incomplete. You are building a URL for a post, but have not provided the post id.

**Comment by miguelgrinberg** [comment]
The stack trace of the error should indicate exactly which `url_for` call is the one with this problem. There is no need to guess, just look at the stack trace and you'll find where you've made a mistake.

**Comment by EryanNeo** [comment]
Dear Miguel, 

I tried looking at it, but Im a bit confused as which one is causing this error.
Below I have attached the error stack. Could you please help lookin at it at quick.
Hope the proficiency you must have, it would be an easy for you to pick the exact error.

```
127.0.0.1 - - [13/Oct/2024 19:15:44] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [13/Oct/2024 19:15:51] "GET /post/110 HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 1498, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 1476, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/main/views.py", line 219, in post
    return render_template('post.html',posts=[post],form=form,pagination=pagination,comments=comments)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/templating.py", line 150, in render_template
    return _render(app, template, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/templating.py", line 131, in _render
    rv = template.render(context)
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/jinja2/environment.py", line 939, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/templates/post.html", line 3, in top-level template code
    {% import "_macros.html" as macros %}
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/templates/base.html", line 1, in top-level template code
    {% extends "bootstrap/base.html" %}
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask_bootstrap/templates/bootstrap/base.html", line 1, in top-level template code
    {% block doc -%}
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask_bootstrap/templates/bootstrap/base.html", line 4, in block 'doc'
    {%- block html %}
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask_bootstrap/templates/bootstrap/base.html", line 20, in block 'html'
    {% block body -%}
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask_bootstrap/templates/bootstrap/base.html", line 23, in block 'body'
    {% block content -%}
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/templates/base.html", line 52, in block 'content'
    {% block page_content %}
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/templates/post.html", line 18, in block 'page_content'
    {{ macros.pagination_widget(pagination, '.post') }}
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/jinja2/runtime.py", line 782, in _invoke
    rv = self._func(*arguments)
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/templates/_macros.html", line 12, in template
    <a href="{{ url_for(endpoint, page = p, **kwargs) }}{{ fragment }}">{{ p }}</a>
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 1084, in url_for
    return self.handle_url_build_error(error, endpoint, values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/flask/app.py", line 1073, in url_for
    rv = url_adapter.build(  # type: ignore[union-attr]
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eryan/Desktop/POPA PIG/Flask/venv/lib/python3.12/site-packages/werkzeug/routing/map.py", line 924, in build
    raise BuildError(endpoint, values, method, self)
werkzeug.routing.exceptions.BuildError: Could not build url for endpoint 'main.post' with values ['page']. Did you forget to specify values ['id']?
127.0.0.1 - - [13/Oct/2024 19:15:51] "GET /post/110?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 304 -
127.0.0.1 - - [13/Oct/2024 19:15:51] "GET /post/110?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 304 -
127.0.0.1 - - [13/Oct/2024 19:15:51] "GET /post/110?__debugger__=yes&cmd=resource&f=console.png&s=CEVlWJL63fQIsCdiNec6 HTTP/1.1" 304 -
127.0.0.1 - - [13/Oct/2024 19:15:55] "GET /post/111 HTTP/1.1" 200 -
127.0.0.1 - - [13/Oct/2024 19:16:57] "POST /post/111 HTTP/1.1" 302 -
127.0.0.1 - - [13/Oct/2024 19:16:57] "GET /post/111 HTTP/1.1" 500 -
```

**Comment by miguelgrinberg** [comment]
All you need to do is search for `url_for` in the stack trace. That gives us this:

```
  File "/home/eryan/Desktop/POPA PIG/Flask/flasky/app/templates/_macros.html", line 12, in template
    <a href="{{ url_for(endpoint, page = p, **kwargs) }}{{ fragment }}">{{ p }}</a>
```

**Comment by EryanNeo** [comment]
Is it something in the macro, where only page is provided. If so, do I also need to provide post Id to the macro function?

**Comment by EryanNeo** [comment]
Dear Miguel,

I performed the changes as mentioned and its working well.
Thank you very much for quick remedy.
Lots of love :-) :-).

Closing this issue.
Thanking you again. Have a great day.

For reference, below is the code snipped changed.

app/mainviews.py:: function:post

return render_template('post.html',posts=[post],form=form,pagination=pagination,comments=comments,id=post.id)

post.html

{{ macros.pagination_widget(pagination, '.post',id=id) }}
